tmail-backend/tmail-third-party/openpaas/src/main/java/com/linagora/tmail/contact/OpenPaasContactsConsumer.java

```diff
@@ -0,0 +1,182 @@
+                    .durable(DURABLE)
+                    .arguments(commonRabbitMQConfiguration.workQueueArgumentsBuilder()
+                    .put("x-dead-letter-exchange", DEAD_LETTER_EXCHANGE)
+                    .put("x-dead-letter-routing-key", EMPTY_ROUTING_KEY)
+                    .build())),
+                sender.bind(BindingSpecification.binding()
+                    .exchange(EXCHANGE_NAME)
+                    .queue(QUEUE_NAME)
+                    .routingKey(EMPTY_ROUTING_KEY)),
+                sender.bind(BindingSpecification.binding()
+                    .exchange(DEAD_LETTER_EXCHANGE)
+                    .queue(DEAD_LETTER_QUEUE)
+                    .routingKey(EMPTY_ROUTING_KEY)))
+            .then()
+            .block();
+
+        consumeContactsDisposable = doConsumeContactMessages();
+    }
+
+    private Disposable doConsumeContactMessages() {
+        return delivery()
+            .flatMap(delivery -> messageConsume(delivery, new String(delivery.getBody(), StandardCharsets.UTF_8)))
+            .subscribe();
+    }
+
+    public Flux<AcknowledgableDelivery> delivery() {
+        return Flux.using(receiverProvider::createReceiver,
+            receiver -> receiver.consumeManualAck(QUEUE_NAME),
+            Receiver::close);
+    }
+
+    private Mono<EmailAddressContact> messageConsume(AcknowledgableDelivery ackDelivery, String messagePayload) {
+        return Mono.just(messagePayload)
+            .map(this::parseContactAddedRabbitMqMessage)
+            .flatMap(this::handleMessage)
+            .doOnSuccess(result -> {
+                LOGGER.warn("Consumed contact successfully '{}'", result);
+                ackDelivery.ack();
+            })
+            .onErrorResume(error -> {
+                LOGGER.error("Error when consume message '{}'", messagePayload, error);
+                ackDelivery.nack(REQUEUE_ON_NACK);
+                return Mono.empty();
+            });
+    }
+
+    private ContactAddedRabbitMqMessage parseContactAddedRabbitMqMessage(String message) {
+        try {
+            return objectMapper.readValue(message, ContactAddedRabbitMqMessage.class);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException("Failed to parse ContactAddedRabbitMqMessage", e);
+        }
+    }
+
+    private Mono<EmailAddressContact> handleMessage(ContactAddedRabbitMqMessage contactAddedMessage) {
+        LOGGER.info("Consumed jCard object message: {}", contactAddedMessage);
+        Optional<ContactFields> contactFieldsOpt = toContactFields(contactAddedMessage.vcard());
+        return openPaasRestClient.retrieveMailAddress(contactAddedMessage.userId())
+            .map(ownerMailAddress -> AccountId.fromUsername(Username.fromMailAddress(ownerMailAddress)))
+            .flatMap(ownerAccountId ->
+                Mono.justOrEmpty(contactFieldsOpt)
+                    .flatMap(contactFields -> doAddContact(ownerAccountId, contactFields)));
```

Comment:
While the type aligns and it likely works, it is not strictly equivalent on the execution side: flatMap is an expensive construct that brings asynchrony into play: it is generally better to keep reactive pipeline simpler, and handle logic resolution upfront if possible (code fragment I proposed and that is now lost in the comment history) but if we want we could save one flatMap with:

```suggestion
            .flatMap(ownerAccountId ->
                contactFieldsOpt.map(contactFields -> doAddContact(ownerAccountId, contactFields))
                   .orElse(Mono.empty());
```

Reply:
Done.